netty/src/main/java/io/grpc/transport/netty/BufferingHttp2ConnectionEncoder.java

```diff
@@ -0,0 +1,313 @@
+    }
+    pendingStream.frames.add(new HeadersFrame(headers, streamDependency, weight, exclusive,
+                             padding, endOfStream, promise));
+    return promise;
+  }
+
+  @Override
+  public ChannelFuture writeRstStream(ChannelHandlerContext ctx, int streamId, long errorCode,
+                                      ChannelPromise promise) {
+    if (existingStream(streamId)) {
+      return super.writeRstStream(ctx, streamId, errorCode, promise);
+    }
+    PendingStream stream = pendingStreams.remove(streamId);
+    if (stream != null) {
+      cancelPendingStream(stream);
+      promise.setSuccess();
+    } else {
+      promise.setFailure(connectionError(PROTOCOL_ERROR, "Stream does not exist %d", streamId));
+    }
+    return promise;
+  }
+
+  @Override
+  public ChannelFuture writeData(ChannelHandlerContext ctx, int streamId, ByteBuf data, int padding,
+                                 boolean endOfStream, ChannelPromise promise) {
+    if (existingStream(streamId)) {
+      return super.writeData(ctx, streamId, data, padding, endOfStream, promise);
+    }
+    PendingStream pendingStream = pendingStreams.get(streamId);
+    if (pendingStream != null) {
+      pendingStream.frames.add(new DataFrame(data, padding, endOfStream, promise));
+    } else {
+      data.release();
+      promise.setFailure(connectionError(PROTOCOL_ERROR, "Stream does not exist %d", streamId));
+    }
+    return promise;
+  }
+
+  @Override
+  public ChannelFuture writeGoAway(ChannelHandlerContext ctx, int lastStreamId, long errorCode,
+                                   ByteBuf debugData, ChannelPromise promise) {
+    cancelPendingStreams();
+    lastStreamId = min(largestCreatedStreamId, lastStreamId);
+    return super.writeGoAway(ctx, lastStreamId, errorCode, debugData, promise);
+  }
+
+  @Override
+  public ChannelFuture writeSettingsAck(ChannelHandlerContext ctx, ChannelPromise promise) {
+    ChannelFuture future = super.writeSettingsAck(ctx, promise);
+    // After having received a SETTINGS frame, the maximum number of concurrent streams
+    // might have changed. So try to create some buffered streams.
+    tryCreatePendingStreams();
+    return future;
+  }
+
+  @Override
+  public void close() {
+    super.close();
+    cancelPendingStreams();
+  }
+
+  private void tryCreatePendingStreams() {
+    while (!pendingStreams.isEmpty() && connection().local().acceptingNewStreams()) {
+      createNextPendingStream();
+    }
+  }
+
+  private void cancelPendingStreams() {
+    while (!pendingStreams.isEmpty()) {
+      PendingStream stream = pendingStreams.pollFirstEntry().getValue();
+      cancelPendingStream(stream);
+    }
+  }
+
+  private void createNextPendingStream() {
+    Map.Entry<Integer, PendingStream> entry = pendingStreams.pollFirstEntry();
+    if (entry == null) {
+      return;
+    }
+    PendingStream pendingStream = entry.getValue();
+    for (Frame frame : pendingStream.frames) {
```

Comment:
Also, you could give `Frame` a `send` method which might get rid of the if/else.